Custom cwd resolution

davidmh opened the pull request:
This feature is meant to open up the cwd resolution to provide support for edge cases around non-writable buffers, like the ones created by vim-fugitive and neogit.

See https://github.com/actionshrimp/direnv.nvim/issues/16 for more details.

In my personal config I'm doing this:

```lua
local git_filetypes = { "git", "fugitive", "fugitiveblame", "NeogitStatus", "NeogitCommitPopup" }

function get_cwd()
	local buf_name = vim.api.nvim_buf_get_name(0)
	if vim.tbl_contains(git_filetypes, vim.o.ft) or vim.tbl_contains(git_filetypes, vim.fs.basename(buf_name)) then
		return vim.fs.root(vim.uv.cwd(), ".git")
	else
		if vim.uv.fs_stat(buf_name) then
			return vim.fs.dirname(buf_name)
		else
			return nil
		end
	end
end

return {
  "davidmh/direnv.nvim",
  branch = "custom-setup",
  opts = {
    async = true,
    get_cwd = get_cwd,
  },
}
```

gn-dave-a reviewed (changes requested):
(posted from wrong account, sorry!)

actionshrimp reviewed (changes requested):
Thanks for this, I definitely think it would be useful to apply to other buffer types. 

I wonder if we could simplify the api slightly by moving the `get_cwd` option to top level, and if unset then defaulting based on 'type', otherwise using the passed function. That way we wouldn't need `get_cwd` in multiple places in the options, and your example config cloud just become:

```
return {
  opts = {
    async = true,
    get_cwd = get_cwd
  },
}
```

What do you think?

(To be honest I could probably do the same with the `autocmd_event` and `autocmd_pattern` options, but I'll think about that separately as it would be nice to avoid breakage).

davidmh commented:
I like it, I'll give it a try this weekend

davidmh committed:
Custom cwd resolution

This feature is meant to open up the cwd resolution to provide support
for edge cases around non-writable buffers, like the ones created by
vim-fugitive and neogit.

See https://github.com/actionshrimp/direnv.nvim/issues/16 for more
details.

actionshrimp reviewed (approved):
Looks great, thanks!
